### PR TITLE
Lu 2685 apply factors sg

### PR DESCRIPTION
**Pull request by kingmushroom**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,44 @@
...
+
+This is the first step in the imputation section of the process. The Wrangler is responsible for checking for strata miss matches, if an anomaly is detected a notification will be created and handed to the BPM.
+
+As imputation will not always run, there needs to be some way of checking if it's an imputation run or not, this is known as check for non responsers. This task has been dealt with inside of the wranger and will bypass the imputation processing and pass the data on if there are no non-responders.
+
+As the correct practice is to seperate out the creation of columns from the method, this wrangler is responsible for creating each questions responding movement column.
+
+Like every wrangler, it is responsible for dealing with sending data to the SQS Queue so that it can move to the next process, it is also responsible for sending data to the BPM.
+
+### Apply Factors Wrangler
```
Comment (GMReynolds) [inline README.md]: Readme includes wrangler info

File: README.md [hunk 2]
```diff
@@ -1 +1,44 @@
...
+
+**Name of Lambda:** imputation_calculate_movement _This Method will be re-named to be inkeeping with the standards soon_
+
+**Intro:** The calculate movement method takes the current year's question value, for each question and subtracts the corresponding previous years question value and then divides the result by the current year's question value **e.g. Question_Movement = (Q106_Current_Year - Q106_Previous_Year) / Q106_Current_Year**
+
+**Inputs:** This method will require all of the Questions columns to be on the data which is being sent to the method, e.g. **Q601,Q602...**. A movement_*question* column should be created for each question in the data wrangler for correct usage of the method. The way the method is written will create the columns if they haven't been created before but for best practice create them in the data wrangler.  
+
+**Outputs:** A Json string which contains all the created movements, saved in the respective movement_*question_name* columns.
+
+### Apply Factors Method
```
Comment (GMReynolds) [inline README.md]: read me includes method description

File: apply_factors_method.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+
+import traceback
+import pandas as pd
+
+
+def _get_traceback(exception):
```
Comment (GMReynolds) [inline apply_factors_method.py]: :param should be :param: - this occurs throughout the doc strinfs

File: README.md (hunk 1 above)
Comment (kingmushroom) [reply README.md]: ….its supposed to?

**Comment by kingmushroom** [comment]
Not included as a comment, but tests didnt work. I've addressed this now following review

**Comment by thomashenson** [comment]
Changes made as requested. Pulled down, tests pass. Good to go.

File: apply_factors_method.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        question_columns = ['Q601_asphalting_sand', 'Q602_building_soft_sand', 'Q603_concreting_sand', 'Q604_bituminous_gravel',
+                            'Q605_concreting_gravel', 'Q606_other_gravel', 'Q607_constructional_fill']
+
+        for question in question_columns:
+            # Loop through each question value, impute based on factor and previous value
+            # then drop the previous value and the imp factor
+            working_dataframe[question] = working_dataframe.apply(lambda x: x['prev_'+question]*x['imputation_factor_'+question], axis=1)
+            working_dataframe = working_dataframe.drop(['prev_'+question, 'imputation_factor_'+question], axis=1)
+
+    except Exception as exc:
```
Comment (bitmonkey) [inline apply_factors_method.py]: This exception block isn't being hit in the `test_apply_factors` test, your tests should cover sad-paths as well as happyt there also looks like there is test code in there e.g

- the `print` statement is a big code smell for me!
- doen't look like any of the import + purge code should be there.

File: apply_factors_wrangler.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+        imputation_run_type = "Imputation complete"
+        message = final_imputed.to_json(orient='records')
+
+        send_output_to_sqs(queue_url,message,sqs_messageid_name,receipt_handle)
+
+        send_sns_message(arn, imputation_run_type, checkpoint)
+
+    except Exception as exc:
+        print(_get_traceback(exc))
```
Comment (bitmonkey) [inline apply_factors_wrangler.py]: More test code? 

There's more print statements and queue purging I'm not sure should be here?

File: apply_factors_wrangler.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                     MessageDeduplicationId=str(random.getrandbits(128))
+                     )
+
+
+def lambda_handler(event, context):
+    try:
+
+        # Set up clients
+        # # S3
+        bucket_name = os.environ['bucket_name']
```
Comment (bitmonkey) [inline apply_factors_wrangler.py]: Need to do the same input validation steps we'd mentioned recently....

`os.environ['some_value']` == bad
`os.environ.get('some_value, None)` + handling None == better
using `marshmallow` to validate your inputs  == happy days

**Comment by bitmonkey** [comment]
@kingmushroom 

Almost there - still seeing a couple of lint errors:

[lint_apply_factors.txt](https://github.com/ONSdigital/es-imputation-sg/files/3440864/lint_apply_factors.txt)

**Comment by kingmushroom** [comment]
Changes made. 
The import os at the top of TestApplyFactors is a false positive. I can't seem to get flake8 to ignore it, but it is used and removing it breaks things.

**Comment by kingmushroom** [comment]
> Changes made.
> The import os at the top of TestApplyFactors is a false positive. I can't seem to get flake8 to ignore it, but it is used and removing it breaks things.

It wasnt, I've removed it. I was looking at the wrong place....

**Comment by bitmonkey** [comment]
@kingmushroom so fixed the remaining lint errors, added some more linting plugins and then had to tweak a few more things. Rather than bounce back again, just did myself....merging.

**Pull request merged**